### DISCUSSION TITLE
I want to get env parameter by using webpack config as function

### DISCUSSION THREAD

**Discussion by mango906** [webpack/webpack#15124] [Q&A] [answered]
## Feature request

**What is the expected behavior?**

I can't get env parameter when i using webpack config as function

**What is motivation or use case for adding/changing the behavior?**

 I'm building a web server with using webpack and webpack-hot-middleware.

When i put `webpack-hot-middleware/client` to every entry like this, my production build bundle file includes `webpack-hot-middleware` package codes

So, I want to put `webpack-hot-middleware/client` to entry array when **only development enviroment** 

Like this

[Referring to this pages to seperate env files](https://webpack.js.org/configuration/mode/)

**webpack.config.js**
```javascript
module.exports = (env, argv) => {
  const config = {};
  const mode = argv.mode || 'development';

  const entry = {
        test: ['./src/script/test.ts'],
    };

  if(mode === 'development'){
    Object.keys(entry).forEach((key) => (entry[key].push('webpack-hot-middleware/client'))
  }

  ...
}
```

**app.js (server.js)**
```javascript
const webpackHotMiddleware = require('webpack-hot-middleware');
const webpackDevMiddleware = require('webpack-dev-middleware');
const webpack = require('webpack');

const webpackConfig = require('./webpack.config');
const compiler = webpack(webpackConfig());

if (node_env === 'local') {
    app.use(webpackDevMiddleware(compiler));
    app.use(webpackHotMiddleware(compiler, { path: '/__webpack_hmr' }));
}
```

**package.json**
```json
"scripts": {
  "start": "webpack --mode=development & npm run app.js"
}
```

Todo pass env, argv parameter to webpack.config.json, i change webpack.config.json Object to function
So i have to pass config like this (using function)
```javascript
const compiler = webpack(webpackConfig());
```

But, I Can't pass parameter to webpack config (Because no parameter in webpackConfig function) and `env`, `argv` parameters are undefined

So, as a temporary measure, Optional Chaining has been used.

```javascript
const mode = argv?.mode || 'development';
```

**How should this be implemented in your opinion?**

I want to be able to receive the env parameter values by using callback function.

Like this

```javascript
const compiler = webpack(((env, argv) => webpackConfig(env, argv))());
```

**Are you willing to work on this yourself?**
yes

**Comment by alexander-akait** [comment, marked as answer]
Sorry, it is impossible, only function inside `webpack.config.js` (or in `--config` argument) can accept these stuff
